crates/networking/p2p/rlpx/connection.rs

```diff
@@ -568,20 +575,56 @@ impl<S: AsyncWrite + AsyncRead + std::marker::Unpin> RLPxConnection<S> {
             Message::NewPooledTransactionHashes(new_pooled_transaction_hashes)
                 if peer_supports_eth =>
             {
-                //TODO(#1415): evaluate keeping track of requests to avoid sending the same twice.
                 let hashes =
                     new_pooled_transaction_hashes.get_transactions_to_request(&self.blockchain)?;
 
-                //TODO(#1416): Evaluate keeping track of the request-id.
-                let request = GetPooledTransactions::new(random(), hashes);
+                // avoid requesting the same hash multiple times
+                let mut hashes_to_request = vec![];
+                for hash in hashes {
+                    if self.get_p2p_transaction(&hash)?.is_none() {
+                        hashes_to_request.push(hash);
+                    }
+                }
+
+                let request_id = random();
+                self.requested_pooled_txs
+                    .insert(request_id, new_pooled_transaction_hashes);
+
+                let request = GetPooledTransactions::new(request_id, hashes_to_request);
                 self.send(Message::GetPooledTransactions(request)).await?;
             }
             Message::GetPooledTransactions(msg) => {
-                let response = msg.handle(&self.blockchain)?;
+                // TODO(#1615): get transactions in batch instead of iterating over them.
+                let txs = msg
+                    .transaction_hashes
+                    .iter()
+                    .map(|hash| self.get_p2p_transaction(hash))
+                    // Return an error in case anything failed.
+                    .collect::<Result<Vec<_>, _>>()?
+                    .into_iter()
+                    // As per the spec, Nones are perfectly acceptable, for example if a transaction was
+                    // taken out of the mempool due to payload building after being advertised.
+                    .flatten()
+                    .collect();
+
+                let response = PooledTransactions::new(msg.id, txs);
                 self.send(Message::PooledTransactions(response)).await?;
             }
             Message::PooledTransactions(msg) if peer_supports_eth => {
                 if self.blockchain.is_synced() {
+                    if let Some(requested) = self.requested_pooled_txs.get(&msg.id) {
+                        if let Err(error) = msg.validate_requested(requested).await {
+                            log_peer_warn(
+                                &self.node,
+                                &format!("disconnected from peer. Reason: {}", error),
+                            );
+                            self.send_disconnect_message(Some(DisconnectReason::SubprotocolError))
+                                .await;
+                            return Err(RLPxError::Disconnected());
```

Comment:
```suggestion
                            return Err(RLPxError::DisconnectSent(
                                DisconnectReason::SubprotocolError,
                            ));
```
Maybe we should even contemplate removing `Disconnected` reason, I remember having created `DisconnectSent` and `DisconnectReceived` for this purpose.